### ISSUE TITLE
recruitment deviations distribution and autocorrelation 

### ISSUE THREAD

**Issue by vbartolino** [gadget-framework/gadget3#211]
I'm looking with interest at this example combining normally distributed and autocorrelated deviations
https://github.com/gadget-framework/gadget3/blob/6757ebd3f24e4a5272e74171eabc77a6b3dcd8c4/tests/test-param_project.R#L14C1-L26C28

It would be useful to have something similar for recruitment but I'm unsure on how to think about it.
Trying to build on previous code, I arrive to something way too complicated, overparametrised, and most importantly impracticable for projections because the lack of correlation between the two distributions. I winder if something like `dnorm` with embedded autocorrelation exist... please tell me if I am talking nonsense

```
gadget3::g3a_renewal_normalparam(imm_stock, run_step=2,
   run_projection = TRUE,
   factor_f =
       g3_formula(y*exp(x*z),
          y = g3_parameterized(name = "ven.rec.scalar", value=100),
          x = g3_param_project(
              "ven.rec.rw",
              by_step = FALSE,
              random = FALSE,
              g3_param_project_rwalk(
                  g3_parameterized("ven.rec_rw_mean", by_stock = FALSE, value=0),
                  g3_parameterized("ven.rec_rw_sd", by_stock = FALSE, value=1000, exponentiate=TRUE)
              )),
          z = g3_param_project(
            "ven.rec",
            by_step = FALSE,
            random = FALSE,
            g3_param_project_dnorm(
              g3_parameterized("ven.rec_mean", by_stock = FALSE, value=4.4),
              g3_parameterized("ven.rec_sd", by_stock = FALSE, value=1000, exponentiate=TRUE)
            ))
          ),
...)
```

**Comment by vbartolino** [comment]
for what is worth, with gadget2, so without the complication of tmb or ambition for any random effect, I remember using `arima.sim`or something along these lines `arima.sim(n=100,list(ar=0.65),rand.gen=rnorm, sd=sqrt(0.2))`

**Comment by lentinj** [comment]
Right, I'm finally getting there :) Whilst meeting @willbutler42 & @bthe we decided that what we need is ``g3_param_project_logar1``. We also mooted a ``g3_param_project_logrwalk``, but that's just logar1 with phi = 0?

Here's the core of what will be ``g3_param_project_logar1``, heavily inspired by the SAM equivalent recruitment function. Any feedback at this stage would be useful.

As with the other projection options, there's 2 functions, an nll half and a simulation half.

The nll generation is ran at the end of the model, and takes a vector (real-space) recruitment rates, logphi & loglevel (an offset):

```r
g3_param_project_nll_logar1 <- g3_native(r = function (var, lstddev, logphi, loglevel) {
        logvar <- log(var) - loglevel
        logphi <- (2.0 / (1.0 + exp(-logphi)) - 1.0);  # Bound between -1 & 1 TODO: Do we need a penalty for this?
        distmean <- 0 - exp(2*lstddev) / 2
        diststddev <- exp(lstddev)

        laglogvar <- c(0, tail(logvar, -1))  # i.e. vector with the previous entry to subtract
        return(-sum(dnorm(
            logvar - logphi * laglogvar,
            distmean,
            diststddev,
            log = TRUE )))
    })
```

The projection function runs in the first projection step, takes a half-finished vector of (real-space) recruitment rates, where the latter entries are all NA, and fills in the gaps:

```r
    g3_param_project_logar1 <- g3_native(r = function (var, lstddev, logphi, loglevel) {
        if (all(is.finite(var))) return(var)
        logvar <- log(var) - loglevel
        logphi <- (2.0 / (1.0 + exp(-logphi)) - 1.0);  # Bound between -1 & 1 TODO: Do we need a penalty for this?

        distmean <- 0 - exp(2*lstddev) / 2
        diststddev <- exp(lstddev)
        for (i in seq_along(logvar)) {
            if (!is.finite(logvar[[i]])) {
                # NB: Start from previous non-projected value if there is one, otherwise 0
                laglogvar <- if (i > 1) logvar[[i - 1]] else 0
                logvar[[i]] <- logphi * laglogvar + rnorm(1, distmean, diststddev)
            }
        }
        var <- exp(logvar + loglevel)
        return(var)
    })
```

Does this seem reasonable? Are any options you'd like to see missing? Is this fundamentally different to what you were hoping for @vbartolino?

More specifically:

* I've added the bounding code for ``logphi`` as seen elsewhere, but I'm not convinced it's necessary, versus setting a lower & upper by default. AFAICS whilst the end result will be ill-defined, it won't be NaN.
* The most natural way to join the projected values & the non-projected values isn't that clear to me. Do we feed the last non-projected value into the AR1 model? Do we start at 0 and let you tweak loglevel so everything lines up? Should loglevel be set to the previous non-projected value?

Annoyingly I don't think the TMB AR1 mechanisms will be that helpful, because we need to apply the bias to the noise distribution, similarly with ``arima.sim``.

**Comment by vbartolino** [comment]
fantastic! as far as I understand, it contains what we need.

This has to be tried to model the deviations from a mean S-R like `bevertonholt` or `hockeystick`, that should be a typical use.

It works in log space which is the way to approach recruitment deviations. I wonder if a non-log version would be of any use for other biological parameters that also might be better capture by an AR1 or that would be unrealistic/unnecessary.

> I've added the bounding code for `logphi` as seen elsewhere

I do not see harm or really a reason for not doing that. I assume, the user can still specify the -1 & +1 boundaries so that a penalty can be applied as those boundaries are approached, right?

> The most natural way to join the projected values & the non-projected values isn't that clear to me...

In all cases we would like continuity of the process from the non-projection to the projection period. I do not see the benefit to tweak the loglevel if that can be embedded in the function. Providing the last non-projected value into the AR1 or using that to set the loglevel seem nearly equivalent to me but there might be statistical reason to prefer one over the other (@bthe?) and it can always be tested.

**Comment by lentinj** [comment]
> This has to be tried to model the deviations from a mean S-R like bevertonholt or hockeystick, that should be a typical use

Yup, that should work fine, as well as directly into renewal. I'll work up my example I'm doing now to include one of the above though.

> I wonder if a non-log version would be of any use for other biological parameters

A non-log version would be very simple to add once this is done. It could use ``AR1``/``arima.sim`` but it could also be done just by simplifying the above a bit.

> I do not see harm or really a reason for not doing that. I assume, the user can still specify the -1 & +1 boundaries so that a penalty can be applied as those boundaries are approached, right?

Yeah, the penalty parameter bounds can still be added, and I'll set them by default. I think these would be required regardless to steer the optimiser.

I guess the main reason to not have a hard bound on top of this is it complicates interpreting the value of ``logphi`` further. It's wasted work too, but this only happens twice in a model run.

**Comment by bthe** [comment]
If you are estimating the correlation then using the bounding code makes sense and then you don't (and probably shouldn't) add bounds to the parameter, as it expects the `logphi` parameters to range from `-Inf` to `Inf`. However would it make sense to have this bounding code as the default, but when the user supplies something different that overwrites the bounding?

**Comment by lentinj** [comment]
Right, I've got something that looks not-terrible in https://github.com/gadget-framework/gadget3/commit/27aed5413570c3cb0701345f27a8f9c9c6038f98. The test is an example of plugging it into hockeyfleet, which is hopefully useful as a starting point if you feel like kicking it's tyres.

Also I've done a bit of work [tidying up the naming of projection parameters](https://github.com/gadget-framework/gadget3/commit/fa9caa7ecd05d2bcfb74f83ae587d18d019f56cc) which were a bit all over the place and added a generic [offset & scale](https://github.com/gadget-framework/gadget3/commit/74acdf8af3a2d93aaa6fbd73f7d715ecfa36e84c).

If this looks reasonable then we could make a non-log version pretty easily, by removing the exp() and the bias.

**Comment by vbartolino** [comment]
I've a couple of functioning implementations of vendace with `g3a_spawn_recruitment_hockeystick` where `r0` is defined with: 

`g3_param_project_logar1()` https://github.com/vbartolino/g3_GoB/tree/f9ab2003d742cc1d25e7c88f6f4e100bd53efbd8/vendace_g3/gadget3/12.03-hs.ar1

`g3_param_project_dnorm()` https://github.com/vbartolino/g3_GoB/tree/f9ab2003d742cc1d25e7c88f6f4e100bd53efbd8/vendace_g3/gadget3/12.03-hs.dnorm2

both with `random = FALSE`.

I've the following issues when doing projections with:
- `_dnorm` the tail of the distribution of the deviations is very long and unrealistically crazy large recruitments occur at times. This is not uncommon (in practice unrelated to g3). Would it make sense to have a threshold? Thoughts and/or suggestions how?
-  `_logar1` whatever parameters I try in the projections for `.lstddev` and `.logphi` the median recruitment drops. Suggestions?

About `g3_param_project_logar1()` I'm also a bit confused why it has only these two parameters and not also a `lmean` parameter. Looking at the test [tests/test-param_project-logar1.R](https://github.com/gadget-framework/gadget3/blob/84bcd93d1ac8e5b6650c18a5929a050b131df0cf/tests/test-param_project-logar1.R) it seems that `r0` would be centered around zero when used in projections which seems problematic. Should `offset` be used to define a sort of mean recruitment level? am I completely off?

**Comment by lentinj** [comment]
@willbutler42 Would an option to use the mean of the last x values help, rather than just the final value? A hard-coded loglevel would also be done, but seems a bit tedious to use.

@vbartolino:
> it seems that r0 would be centered around zero when used in projections which seems problematic

I *think* this is the same question, when projecting we start at the final non-projection value (loglevel), and use this to offset the projections.

(the random noise also has a mean that's hard-coded to 0, but I don't think this is what you want).

But I'll add an option to hard-code ``loglevel`` now, then have a look at the other parts to your comment.

**Comment by willbutler42** [comment]
@lentinj Yep, that would help thanks

**Comment by vbartolino** [comment]
> @willbutler42 Would an option to use the mean of the last x values help, rather than just the final value? A hard-coded loglevel would also be done, but seems a bit tedious to use.

leaving the possibility to specify ourselves from which years to get the `loglevel` would be ideal. Would it allow something like
```
params.proj <- params.out %>%
  g3_init_val('project_years', 12) %>%
  g3_init_val('ven.rec.*.loglevel',
              value=params.out %>% filter(switch %in% paste0("ven.rec.",2000:(tyr-1))) %>% .$value %>% as.numeric() %>% mean()
```

> I think this is the same question, when projecting we start at the final non-projection value (loglevel), and use this to offset the projections.

now I also understand what we man by loglevel :)
yes, it's the same question

**Comment by willbutler42** [comment]
>     * `_logar1` whatever parameters I try in the projections for `.lstddev` and `.logphi` the median recruitment drops. Suggestions?
> 

I had this with blue ling too, possibly to do with the default values? 
```
lstddev <- 1
noisemean <- 0 - exp(2*lstddev) / 2
noisestddev <- exp(lstddev)
rnorm(1000, noisemean, noisestddev) |> mean()
```

**Comment by lentinj** [comment]
Right, https://github.com/gadget-framework/gadget3/commit/03a8b668693904ad36254936723d3c1a61c2e05f adds a loglevel parameter. This defaults to "-1" (i.e. use the last value), but if positive then will use that instead.

Obviously the next step is to support -n to use the last n values, but I've not done that just yet.

**Comment by lentinj** [comment]
> _dnorm the tail of the distribution of the deviations is very long and unrealistically crazy large recruitments occur at times. This is not uncommon (in practice unrelated to g3). Would it make sense to have a threshold?

You could continually call rnorm() until all values are within the given bounds, but how you'd generate the corresponding distribution for likelihood, I'm not sure.

Seems likely there's a standard approach I don't know about though.

**Comment by lentinj** [comment]
@willbutler42 the above supports loglevel -x to use mean of last x values now. If your x is larger than what's available, then it falls back to 0.

**Comment by vbartolino** [comment]
337e383852d2f2bd1b9442b2f9e6efff7a8645a9 gives much more control on the level of projected recruitment, I've just tried.

**Comment by lentinj** [comment]
https://github.com/gadget-framework/gadget3/commit/2479f05149ae130ca36199e4d50d7e1dc658fcb2 also implements ``g3_param_project_ar1`` whilst I was here. Using the native implementations might be a bit faster, but can come back and splice them in if needed.
